lib/xgu/xgu.h

```diff
@@ -0,0 +1,208 @@
+  XGU_TEXCOORD1_ARRAY = NV2A_VERTEX_ATTR_TEXTURE1,
+  XGU_TEXCOORD2_ARRAY = NV2A_VERTEX_ATTR_TEXTURE2,
+  XGU_TEXCOORD3_ARRAY = NV2A_VERTEX_ATTR_TEXTURE3
+} XguVertexArray;
+
+static inline
+DWORD* push_command(DWORD* p, DWORD command, int nparams) {
+    pb_push(p++, command, nparams);
+    return p;
+}
+
+static inline
+DWORD* push_parameter(DWORD* p, DWORD parameter) {
+    *p++ = parameter;
+    return p;
+}
+
+static inline
+DWORD* push_boolean(DWORD* p, bool enabled) {
+  return push_parameter(p, enabled ? 1 : 0);
+}
+
+static inline
+DWORD* push_command_boolean(DWORD* p, DWORD command, bool enabled) {
+  p = push_command(p, command, 1);
+  p = push_boolean(p, enabled);
+  return p;
+}
+
+static inline
+DWORD* push_float(DWORD* p, float f) {
+  return push_parameter(p, *(uint32_t*)&f);
+}
+
+static inline
+DWORD* push_floats(DWORD* p, float* f, unsigned int count) {
+  for(unsigned int i = 0; i < count; i++) {
+    p = push_float(p, f[i]);
+  }
+  return p;
+}
+
+static inline
+DWORD* push_matrix2x2(DWORD* p, float m[2*2]) {
+  return push_floats(p, m, 2*2);
+}
+
+static inline
+DWORD* push_matrix4x4(DWORD* p, float m[4*4]) {
+  return push_floats(p, m, 4*4);
+}
+
+static inline
+DWORD* push_command_matrix2x2(DWORD* p, DWORD command, float m[2*2]) {
+  p = push_command(p, command, 2*2);
+  p = push_matrix2x2(p, m);
+  return p;
+}
+
+static inline
+DWORD* push_command_matrix4x4(DWORD* p, DWORD command, float m[4*4]) {
+  p = push_command(p, command, 4*4);
+  p = push_matrix4x4(p, m);
+  return p;
+}
+
+static inline
+DWORD* push_command_parameter(DWORD* p, DWORD command, DWORD parameter) {
+  p = push_command(p, command, 1);
+  p = push_parameter(p, parameter);
+  return p;
+}
+
+inline
+DWORD* xgu_begin(DWORD* p, XguPrimitiveType type) {
+
+  // Force people to use xgu_end instead
+  assert(type != NV097_SET_BEGIN_END_OP_END);
+
+  return push_command_parameter(p, NV097_SET_BEGIN_END, type);
+}
+
+inline
+DWORD* xgu_end(DWORD* p) {
+  return push_command_parameter(p, NV097_SET_BEGIN_END, NV097_SET_BEGIN_END_OP_END);
+}
+
+inline
+DWORD* xgu_clear_color(DWORD* p, DWORD color) {
+    p = push_command_parameter(p, NV097_SET_COLOR_CLEAR_VALUE, color);
+    return p;
+}
+
+inline
+DWORD* xgu_clear(DWORD* p) {
+    /* Should probably do some magic related to 
+     * NV20_TCL_PRIMITIVE_3D_CLEAR_VALUE_HORIZ and
+     * NV20_TCL_PRIMITIVE_3D_CLEAR_VALUE_DEPTH
+     */
+    return p;
+}
+
+inline
+DWORD* xgu_set_object(DWORD* p, uint32_t instance) {
+    p = push_command(p, NV097_SET_OBJECT, 1);
+    p = push_parameter(p, instance);
+    return p;
+}
+
+inline
+DWORD* xgu_set_model_view_matrix(DWORD* p, float m[4*4]) {
+    return push_command_matrix4x4(p, NV097_SET_MODEL_VIEW_MATRIX, m);
+}
+
+inline
+DWORD* xgu_matrix_mode_projection(DWORD* p, float m[4*4]) {
+    return push_command_matrix4x4(p, NV097_SET_PROJECTION_MATRIX, m);
+}
+
+inline
+DWORD* xgu_set_texture_matrix(DWORD* p, unsigned int index, float m[2*2]) {
```

Comment:
index = texture slot / stage, name should be improved to reflect this

Reply:
Better?

Reply:
Yeah - better; we still might have to refactor eventually, because the names are rather ambigious and there's other kind slots/stages, too